File: LWidget/src/LButton.cpp
#include "LButton.h"
#include <QPainterPath>
using namespace ljz;

void LSwitchButton::StateInfo::reSize(double radius)
{
	if (this->buttonPixmap.isNull())
		this->circlePixmap = QPixmap();
	else
	{
		const int size = static_cast<int>(radius * 2);
		if (this->circlePixmap.width() == size && this->circlePixmap.height() == size && !this->circlePixmap.isNull())
			return;
		QPixmap pixmap(size, size);
		QPixmap scaled = this->buttonPixmap.scaled(pixmap.size(), Qt::IgnoreAspectRatio, Qt::SmoothTransformation);
		pixmap.fill(Qt::transparent);
		QPainter painter(&pixmap);
		painter.setRenderHint(QPainter::Antialiasing);
		QPainterPath path;
		path.addEllipse(0, 0, size, size);
		painter.setClipPath(path);
		painter.drawPixmap(0, 0, scaled);
		this->circlePixmap = pixmap;
	}
}

LSwitchButton::StateInfo::StateInfo(State state)
{
	switch (state)
	{
	case State::On:
	{
		this->text = "On";
		this->buttonPixmap = QPixmap();
		this->backgroundColor = Qt::white;
		this->textColor = Qt::black;
		break;
	}
	case State::Off:
	{
		this->text = "Off";
		this->buttonPixmap = QPixmap();
		this->backgroundColor = Qt::black;
		this->textColor = Qt::white;
		break;
	}
	default:
		break;
	}
}

LSwitchButton::StateInfo::StateInfo(QString text, QPixmap buttonPixmap, const QColor& backgroundColor,
	const QColor& textColor) : text(std::move(text)), buttonPixmap(std::move(buttonPixmap)), backgroundColor(backgroundColor), textColor(textColor)
{}

void LSwitchButton::setStateInfo(State state, const StateInfo& stateInfo)
{
	this->_stateInfoMap.insert(state, stateInfo);
}

LSwitchButton::StateInfo LSwitchButton::stateInfo(State state) const
{
	return this->_stateInfoMap.value(state);
}

void LSwitchButton::setState(State state)
{
	this->setChecked(state == State::On);
}

LSwitchButton::State LSwitchButton::state() const
{
	return this->isChecked() ? State::On : State::Off;
}

void LSwitchButton::setAnimationDuration(int duration) const
{
	this->_animation->setDuration(duration);
}

int LSwitchButton::animationDuration() const
{
	return this->_animation->duration();
}

void LSwitchButton::setOpenAnimation(bool isOpenAnimation)
{
	this->_isOpenAnimation = isOpenAnimation;
}

bool LSwitchButton::isOpenAnimation() const
{
	return this->_isOpenAnimation;
}

LSwitchButton::LSwitchButton(QWidget* parent)
	:QAbstractButton(parent)
{
	this->setMouseTracking(true);
	this->setCheckable(true);
	connect(this, &QAbstractButton::clicked, this, &LSwitchButton::changeState);
	this->_animation->setDuration(100);
	connect(this->_animation, &QPropertyAnimation::finished, [this]()
		{
			this->updateValueRange();
			this->setPosition(this->_animation->startValue().toPointF());
		});
}

LSwitchButton::~LSwitchButton() = default;

void LSwitchButton::paintEvent(QPaintEvent* e)
{
	QPainter painter(this);
	painter.setRenderHint(QPainter::Antialiasing);
	StateInfo info = this->_stateInfoMap.value(this->state());
	info.reSize(_radius);
	bool isAnimationRunning = this->_animation->state() == QAbstractAnimation::Running;
	double radius = this->calRadius(15) * 1.7;

	// 绘制背景
	painter.setPen(Qt::NoPen);
	painter.setBrush(info.backgroundColor);
	painter.drawRoundedRect(_buttonRect, radius, radius);

	// 如果动画运行中,绘制动画未移动到的区域背景
	if (isAnimationRunning)
	{
		double width = this->state() == State::On ? _buttonRect.bottomRight().x() - _position.x() : _position.x() - _buttonRect.topLeft().x();
		painter.setBrush(info.textColor);
		QRectF rect = QRectF(0, _buttonRect.top(), width, _buttonRect.height());
		if (this->state() == State::On)
			rect.moveLeft(_position.x());
		else
			rect.moveRight(_position.x());
		painter.drawRoundedRect(rect, radius, radius);
	}

	// 绘制按钮
	painter.setPen(info.textColor);
	if (info.circlePixmap.isNull())
	{
		painter.setBrush(isAnimationRunning ? info.backgroundColor : info.textColor);
		painter.drawEllipse(_position, _radius, _radius);
	}
	else
	{
		QRect pixmapRect = info.circlePixmap.rect();
		pixmapRect.moveCenter(_position.toPoint());
		painter.drawPixmap(pixmapRect, info.circlePixmap);
	}
	// 绘制文字
	QFontMetrics metrics(this->font());
	QRect textRect = metrics.boundingRect(info.text);
	textRect.moveCenter(_buttonRect.center().toPoint());
	painter.drawText(textRect, Qt::AlignCenter, info.text);
	// 绘制边框
	painter.setPen(info.textColor);
	painter.setBrush(Qt::NoBrush);
	painter.drawRoundedRect(_buttonRect, radius, radius);
	// 如果禁用，绘制遮罩
	if (!isEnabled())
	{
		painter.setBrush(QColor(128, 128, 128, 128));
		painter.setPen(Qt::NoPen);
		painter.drawRoundedRect(_buttonRect, radius, radius);
	}
}

void LSwitchButton::updateValueRange() const
{
	if (this->state() == State::On)
	{
		this->_animation->setStartValue(_right);
		this->_animation->setEndValue(_left);
	}
	else
	{
		this->_animation->setStartValue(_left);
		this->_animation->setEndValue(_right);
	}
}

void LSwitchButton::resizeEvent(QResizeEvent* event)
{
	QAbstractButton::resizeEvent(event);

	this->setRadius(this->calRadius(12));
	this->_left = QPointF(_radius + 5, this->height() / 2.0);
	this->_right = QPointF(this->width() - _radius - 5, this->height() / 2.0);
	this->_buttonRect = QRectF(this->rect().x() + 2, this->rect().y() + 2, this->rect().width() - 4, this->rect().height() - 4);
	this->updateValueRange();
	this->setPosition(this->_animation->startValue().toPointF());
}

QSize LSwitchButton::sizeHint() const
{
	return { 60, 23 };
}

void LSwitchButton::enterEvent(QEnterEvent* event)
{
	QAbstractButton::enterEvent(event);
}

void LSwitchButton::leaveEvent(QEvent* event)
{
	QAbstractButton::leaveEvent(event);
	if (this->isEnabled())
		this->setRadius(this->calRadius(12));
}

bool LSwitchButton::hitButton(const QPoint& pos) const
{
	return _buttonRect.contains(pos);
}

void LSwitchButton::mouseMoveEvent(QMouseEvent* event)
{
	QAbstractButton::mouseMoveEvent(event);
	if (this->isEnabled())
	{
		if (_buttonRect.contains(event->position()))
			this->setRadius(this->calRadius(10));
		else
			this->setRadius(this->calRadius(12));
	}
}

void LSwitchButton::setRadius(double radius)
{
	this->_radius = radius;
	update();
}

double LSwitchButton::calRadius(double num) const
{
	return qMin((this->width() - num) / 2.0, (this->height() - num) / 2.0);
}

void LSwitchButton::changeState(bool checked)
{
	State state = checked ? State::On : State::Off;
	emit stateChanged(state, this->_stateInfoMap.value(state));
	if (this->_isOpenAnimation)
		this->_animation->start();
	else
		this->setPosition(state == State::On ? _right : _left);
}

void LSwitchButton::setPosition(QPointF newValue)
{
	this->_position = newValue;
	update();
}

File: Python/LWidget/LButton.py
from enum import Enum
from PySide6.QtCore import Signal, QPointF, QRectF, Qt, QSize, QPropertyAnimation, Property
from PySide6.QtGui import QPainter, QPixmap, QPainterPath, QFontMetrics, QColor
from PySide6.QtWidgets import QPushButton, QAbstractButton


class LPixmapButton(QPushButton):
    def __init__(self, parent=None):
        super().__init__(parent)
        self._pixmap = QPixmap()
        self._drawPixmap = QPixmap()

    def setPixmap(self, pixmap: QPixmap):
        self._pixmap = pixmap
        self._drawPixmap = pixmap
        mask = self._drawPixmap.createMaskFromColor(Qt.GlobalColor.transparent)
        self.setFixedSize(self._drawPixmap.size())
        self.setMask(mask)

    def pixmap(self):
        return self._pixmap

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        painter.drawPixmap(0, 0, self._drawPixmap)

    def mousePressEvent(self, event):
        if not self.mask().contains(event.pos()):
            event.ignore()
        else:
            super().mousePressEvent(event)

    def enterEvent(self, event):
        super().enterEvent(event)
        self._drawPixmap = self.pixmap().scaled(self.pixmap().size() * 1.05, Qt.AspectRatioMode.ASpIgnoreAspectRatio,
                                                Qt.TransformationMode.SmoothTransformation)
        mask = self._drawPixmap.createMaskFromColor(Qt.GlobalColor.transparent)
        self.setFixedSize(self._drawPixmap.size())
        self.setMask(mask)

    def leaveEvent(self, event):
        super().leaveEvent(event)
        self.setPixmap(self.pixmap())


class LSwitchButton(QAbstractButton):
    stateChanged = Signal(object, object)

    class State(Enum):
        Off = 0
        On = 1

    class StateInfo:
        def __init__(self, state=None):
            if state is None:
                self.text = ""
                self.buttonPixmap = QPixmap()
                self.backgroundColor = QColor()
                self.textColor = QColor()
                self.circlePixmap = QPixmap()
            else:
                if state == LSwitchButton.State.On:
                    self.text = "On"
                    self.buttonPixmap = QPixmap()
                    self.backgroundColor = Qt.GlobalColor.white
                    self.textColor = Qt.GlobalColor.black
                else:
                    self.text = "Off"
                    self.buttonPixmap = QPixmap()
                    self.backgroundColor = Qt.GlobalColor.black
                    self.textColor = Qt.GlobalColor.white

        def reSize(self, radius):
            if self.buttonPixmap.isNull():
                self.circlePixmap = QPixmap()
            else:
                size = int(radius * 2)
                if self.circlePixmap.width() == size and self.circlePixmap.height() == size and not self.circlePixmap.isNull():
                    return
                pixmap = QPixmap(size, size)
                scaled = self.buttonPixmap.scaled(pixmap.size(), Qt.AspectRatioMode.IgnoreAspectRatio,
                                                  Qt.TransformationMode.SmoothTransformation)
                pixmap.fill(Qt.GlobalColor.transparent)
                painter = QPainter(pixmap)
                painter.setRenderHint(QPainter.RenderHint.Antialiasing)
                path = QPainterPath()
                path.addEllipse(0, 0, size, size)
                painter.setClipPath(path)
                painter.drawPixmap(0, 0, scaled)
                self.circlePixmap = pixmap

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setMouseTracking(True)
        self.setCheckable(True)
        self.clicked.connect(self.changeState)
        self._position = QPointF(0, 0)
        self._left = QPointF(0, 0)
        self._right = QPointF(0, 0)
        self._radius = 0
        self._buttonRect = QRectF(0, 0, 0, 0)
        self._isOpenAnimation = True
        self._animation = QPropertyAnimation(self, b'position')
        self._animation.setDuration(100)
        self._animation.finished.connect(self.updateValueRange)
        self._stateInfoMap = {self.State.Off: self.StateInfo(self.State.Off),
                              self.State.On: self.StateInfo(self.State.On)}

    def setStateInfo(self, state, stateInfo):
        self._stateInfoMap[state] = stateInfo

    def stateInfo(self, state):
        return self._stateInfoMap[state]

    def setState(self, state):
        self.setChecked(state == self.State.On)

    @Property(State, fset=setState)
    def state(self):
        return self.State.On if self.isChecked() else self.State.Off

    def setAnimationDuration(self, duration):
        self._animation.setDuration(duration)

    def animationDuration(self):
        return self._animation.duration()

    def setOpenAnimation(self, isOpenAnimation):
        self._isOpenAnimation = isOpenAnimation

    def isOpenAnimation(self):
        return self._isOpenAnimation

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        stateInfo = self._stateInfoMap[self.state]
        stateInfo.reSize(self._radius)
        isAnimationRunning = self._animation.state() == QPropertyAnimation.State.Running
        radius = self.calRadius(15) * 1.7

        # Draw background
        painter.setPen(Qt.NoPen)
        painter.setBrush(stateInfo.backgroundColor)
        painter.drawRoundedRect(self._buttonRect, radius, radius)

        # If animation is running, draw the background of the area not moved by the animation
        if isAnimationRunning:
            width = self._buttonRect.bottomRight().x() - self._position.x() if self.state == self.State.On else self._position.x() - self._buttonRect.topLeft().x()
            painter.setBrush(stateInfo.textColor)
            rect = QRectF(0, self._buttonRect.top(), width, self._buttonRect.height())
            if self.state == self.State.On:
                rect.moveLeft(self._position.x())
            else:
                rect.moveRight(self._position.x())
            painter.drawRoundedRect(rect, radius, radius)

        # Draw button
        painter.setPen(stateInfo.textColor)
        if stateInfo.circlePixmap.isNull():
            painter.setBrush(Qt.GlobalColor.transparent if isAnimationRunning else stateInfo.textColor)
            painter.drawEllipse(self._position, self._radius, self._radius)
        else:
            pixmapRect = stateInfo.circlePixmap.rect()
            pixmapRect.moveCenter(self._position.toPoint())
            painter.drawPixmap(pixmapRect, stateInfo.circlePixmap)

        # Draw text
        fontMetrics = QFontMetrics(self.font())
        textRect = fontMetrics.boundingRect(stateInfo.text)
        textRect.moveCenter(self._buttonRect.center().toPoint())
        painter.drawText(textRect, Qt.AlignmentFlag.AlignCenter, stateInfo.text)

        # Draw border
        painter.setPen(stateInfo.textColor)
        painter.setBrush(Qt.BrushStyle.NoBrush)
        painter.drawRoundedRect(self._buttonRect, radius, radius)

        # If disabled, draw mask
        if not self.isEnabled():
            painter.setBrush(QColor(128, 128, 128, 128))
            painter.setPen(Qt.PenStyle.NoPen)
            painter.drawRoundedRect(self._buttonRect, radius, radius)

    def updateValueRange(self):
        if self.state == self.State.On:
            self._animation.setStartValue(self._right)
            self._animation.setEndValue(self._left)
        else:
            self._animation.setStartValue(self._left)
            self._animation.setEndValue(self._right)

    def resizeEvent(self, event):
        super().resizeEvent(event)
        self.setRadius(self.calRadius(12))
        self._left = QPointF(self._radius + 5, self.height() / 2.0)
        self._right = QPointF(self.width() - self._radius - 5, self.height() / 2.0)
        self._buttonRect = QRectF(self.rect().x() + 2, self.rect().y() + 2, self.rect().width() - 4,
                                  self.rect().height() - 4)
        self.updateValueRange()
        self.setPosition(self._animation.startValue())

    def sizeHint(self):
        return QSize(60, 23)

    def enterEvent(self, event):
        super().enterEvent(event)

    def leaveEvent(self, event):
        super().leaveEvent(event)
        if self.isEnabled():
            self.setRadius(self.calRadius(12))

    def hitButton(self, pos):
        return self._buttonRect.contains(pos)

    def mouseMoveEvent(self, event):
        super().mouseMoveEvent(event)
        if self.isEnabled():
            if self._buttonRect.contains(event.position()):
                self.setRadius(self.calRadius(10))
            else:
                self.setRadius(self.calRadius(12))

    def setRadius(self, radius):
        self._radius = radius
        self.update()

    def calRadius(self, num):
        return min((self.width() - num) / 2.0, (self.height() - num) / 2.0)

    def changeState(self, checked):
        state = self.State.On if checked else self.State.Off
        self.stateChanged.emit(state, self._stateInfoMap[state])
        if self._isOpenAnimation:
            self._animation.start()
        else:
            self.setPosition(self._right if state == self.State.On else self._left)

    def setPosition(self, newValue):
        self._position = newValue
        self.update()

    @Property(QPointF, fset=setPosition)
    def position(self):
        return self._position
